Fixed some docs in std.algorithm

JackStouffer opened the pull request:
* fixed two spelling mistakes
* Added missing params and returns
* removed in doc examples (one of which didn't even compile) for unit test examples
* fixed examples for `reduce` not being included

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -889,6 +884,7 @@ template each(alias pred = "a")
 }
 
+///
 unittest
```
Ddoc'd unittest examples need to be vetted to make sure they make sense to the end user. Comments like "this is not unittestable", etc., don't make any sense from the user's POV and should not be included in ddoc'd unittests.  They should either be left as undocumented unittests, or edited so that it makes sense to someone reading the docs.

JackStouffer replied:
Fixed

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -788,24 +801,6 @@ Params:
-        .filter!(de => de.timeLastModified < cutoff)
-        .each!remove();
-}
----
```
I did not see this in the ddoc'd unittests. Please don't remove in-doc examples that aren't also present in unittest blocks unless something equivalent / similar is put in its place.

JackStouffer replied:
This code sample does not compile, and the other examples are much better at explaining its uses anyway so I didn't try to fix it.

quickfur replied:
Fair enough.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -788,24 +801,6 @@ Params:
----
-arr.each!((ref a) => a++);
-arr.each!"a++";
----
```
This example probably should be reflected in a ddoc'd unittest too. There's one in the current unittest but it's rather obscure; perhaps this particular example should be expanded to be compilable, and put in its own unittest block. Don't forget to include the sentence above it, as it is explaining an important feature of this function that the user should know, so it should either be in the ddoc comment before the unittest block, or in a comment in the unittest itself.

JackStouffer replied:
Fixed

quickfur commented:
LGTM. Thanks!

quickfur commented on std/algorithm/mutation.d:
```diff
@@ -2350,6 +2372,13 @@ Returns a tuple containing the remainder portions of $(D r1) and $(D
+
+Params:
+    r1 = an input range with swapable elements
+    r2 = an input range with swapable elements
```
Typo: `swapable` -> `swappable`

JackStouffer commented:
Fixed. I also updated more docs in algorithm.mutation, adding more params and returns and removed two comments examples to unit tests, one of which had no unit tests to begin with 0_0!

quickfur commented on std/algorithm/mutation.d:
```diff
@@ -755,6 +756,19 @@ void initializeAll(Range)(Range range)
+{
+    struct S
+    {
+        int a;
```
How about testing the case where the default value of `a` is non-zero?

quickfur commented on std/algorithm/mutation.d:
```diff
@@ -2419,3 +2462,11 @@ void uninitializedFill(Range, Value)(Range range, Value value)
+///
+@safe nothrow unittest
+{
+    auto s = new int[5];
```
Are you sure this is equivalent to the original example? IIRC the GC will initialize memory, but `malloc` will not, so there's a difference here.

JackStouffer commented:
Fixed comments.

atilaneves commented:
LGTM

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -425,6 +431,13 @@ returns a range of which elements are obtained by applying $(D fun(a))
+    r = an input range
+
+Returns:
+    r with each fun applied to all the elements
```
This does not make it clear what exactly is returned when more than one function is specified in `fun`. For example, what does `map!("a+a", "a*a")` return?

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -788,24 +801,6 @@ Params:
-arr.each!"a++";
----
-
 If no predicate is specified, $(D each) will default to doing nothing
```
This paragraph is now attached to the description of `r` under `Params`. It doesn't seem to fit there. Maybe it should be moved before the `Params` section?

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -889,35 +884,36 @@ template each(alias pred = "a")
     assert(m.empty);
 
-    // in-place mutation with index
+    // Indexes are also available for in place mutations
```
Nitpick: "in-place" should be hyphenated.

quickfur commented on std/algorithm/mutation.d:
```diff
@@ -755,6 +756,21 @@ void initializeAll(Range)(Range range)
     range[] = T.init;
 }
 
+///
```
This unittest isn't showing up in the generated docs, because the second overload of `initializeAll` is preceded by `// ditto` (missing one `/`), instead of `/// ditto`, so ddoc fails to recognize it as part of the previous declaration.

quickfur commented on std/algorithm/mutation.d:
```diff
@@ -1539,6 +1555,14 @@ calls to $(D range.popFront).)  $(LI Otherwise, elements are moved
 cases.))
+
+Params:
+    s = a SwapStrategy
```
I think it's clear from the declaration itself that `s` is a `SwapStrategy`. The question is, what does that mean in the context of this function?

quickfur commented on std/algorithm/mutation.d:
```diff
@@ -1896,6 +1930,9 @@ if (isRandomAccessRange!Range && hasLength!Range)
 Reverses $(D r) in-place, where $(D r) is a narrow string (having
 elements of type $(D char) or $(D wchar)). UTF sequences consisting of
 multiple code units are preserved properly.
```
This doesn't necessarily have to be addressed in this PR, but this function does *not* properly handle reversal of sequences of combining characters. For example, reversing `ba\u0301d` ("bád") will result in `d\u0301ab` ("d́ab") instead of `da\u0301b` ("dáb"), which is likely *not* what the user would expect. This probably should be noted in a `Bugs:` section.

quickfur commented:
Sorry for the late feedback, been busy at work.

Next time, consider splitting such PRs into multiple self-contained PRs, so that nitpicks in one small part won't hold back other changes that are already ready to merge. Probably one PR per function or group of related functions would work well. As it stands, most of this PR is already good to merge but can't because some other parts still need a bit of work, which is a pity, since many of these fixes ought to be checked in already.

JackStouffer commented:
Fixed all issues. Also, next time I will break up my PRs into smaller chunks.

DmitryOlshansky commented on std/algorithm/mutation.d:
```diff
@@ -2419,3 +2470,13 @@ void uninitializedFill(Range, Value)(Range range, Value value)
+{
+    import core.stdc.stdlib : malloc;
+
+    auto s = (cast(int*) malloc(5 * int.sizeof))[0 .. 5];
```
This leaks.. As an example it's *e-hm* fine. But unittest should absolutely do `scope(exit) free(s.ptr);`

JackStouffer commented:
@DmitryOlshansky fixed

DmitryOlshansky commented:
LGTM then

DmitryOlshansky commented:
> @DmitryOlshansky fixed

Rebase to nudge the auto-tester...

JackStouffer commented:
@DmitryOlshansky rebased, seems to be passing now.

DmitryOlshansky commented:
Auto-merge toggled on

JackStouffer commented:
Thanks!
